diff --git a/keepalived/check/check_daemon.c b/keepalived/check/check_daemon.c
index 432949c9..81f07e53 100644
--- a/keepalived/check/check_daemon.c
+++ b/keepalived/check/check_daemon.c
@@ -108,6 +108,8 @@ stop_check(int status)
 
 	closelog();
 
+	FREE(config_id);
+
 #ifndef _MEM_CHECK_LOG_
 	FREE_PTR(check_syslog_ident);
 #else
diff --git a/keepalived/core/main.c b/keepalived/core/main.c
index d9e4125b..e16a7d00 100644
--- a/keepalived/core/main.c
+++ b/keepalived/core/main.c
@@ -36,6 +36,7 @@
 #include "main.h"
 #include "config.h"
 #include "git-commit.h"
+#include "utils.h"
 #include "signals.h"
 #include "pidfile.h"
 #include "bitops.h"
@@ -980,8 +981,15 @@ keepalived_main(int argc, char **argv)
 	}
 
 	/* daemonize process */
-	if (!__test_bit(DONT_FORK_BIT, &debug))
-		xdaemon(0, 0, 0);
+	if (!__test_bit(DONT_FORK_BIT, &debug)) {
+		close_std_fd();
+		if (xdaemon(false, false, true) > 0) {
+			closelog();
+			FREE(config_id);
+			FREE(orig_core_dump_pattern);
+			exit(0);
+		}
+	}
 
 	/* Set file creation mask */
 	umask(0);
@@ -1043,6 +1051,8 @@ end:
 
 	closelog();
 
+	FREE(config_id);
+
 #ifndef _MEM_CHECK_LOG_
 	FREE_PTR(syslog_ident);
 #else
diff --git a/keepalived/vrrp/vrrp_daemon.c b/keepalived/vrrp/vrrp_daemon.c
index 677b0ee8..adf26add 100644
--- a/keepalived/vrrp/vrrp_daemon.c
+++ b/keepalived/vrrp/vrrp_daemon.c
@@ -179,6 +179,8 @@ stop_vrrp(int status)
 
 	closelog();
 
+	FREE(config_id);
+
 #ifndef _MEM_CHECK_LOG_
 	FREE_PTR(vrrp_syslog_ident);
 #else
diff --git a/lib/parser.c b/lib/parser.c
index 6ed98374..e6d2f233 100644
--- a/lib/parser.c
+++ b/lib/parser.c
@@ -33,6 +33,7 @@
 #include <sys/types.h>
 #include <fcntl.h>
 #include <stdbool.h>
+#include <pwd.h>
 
 #include "parser.h"
 #include "memory.h"
@@ -755,6 +756,9 @@ init_data(const char *conf_file, vector_t * (*init_keywords) (void))
 	read_conf_file(conf_file);
 	unregister_null_strvec_handler();
 
+	/* Close the password database if it was opened */
+	endpwent();
+
 	free_keywords(keywords);
 	clear_rt_names();
 }
