diff --git a/keepalived/core/daemon.c b/keepalived/core/daemon.c
index a8ec37ec..8dc9a5b9 100644
--- a/keepalived/core/daemon.c
+++ b/keepalived/core/daemon.c
@@ -29,7 +29,7 @@
 
 /* Daemonization function coming from zebra source code */
 pid_t
-xdaemon(int nochdir, int noclose, int exitflag)
+xdaemon(bool nochdir, bool noclose, bool exitflag)
 {
 	pid_t pid;
 	int ret;
@@ -65,8 +65,8 @@ xdaemon(int nochdir, int noclose, int exitflag)
 	}
 
 	/* File descriptor close. */
-	if (!noclose)
-		set_std_fd(true);
+//	if (!noclose)
+//		set_std_fd(true);
 
 	return 0;
 }
diff --git a/keepalived/include/daemon.h b/keepalived/include/daemon.h
index 43bc5b4a..bd727ce6 100644
--- a/keepalived/include/daemon.h
+++ b/keepalived/include/daemon.h
@@ -32,8 +32,9 @@
 #include <unistd.h>
 #include <errno.h>
 #include <fcntl.h>
+#include <stdbool.h>
 
 /* prototype */
-extern pid_t xdaemon(int, int, int);
+extern pid_t xdaemon(bool, bool, bool);
 
 #endif
