diff --git a/lib/notify.c b/lib/notify.c
index 9c371986..ab5398fd 100644
--- a/lib/notify.c
+++ b/lib/notify.c
@@ -375,9 +378,10 @@ system_call_script(thread_master_t *m, int (*func) (thread_t *), void * arg, uns
 void
 script_killall(thread_master_t *m, int signo)
 {
-	sigset_t old_set, child_wait;
 	thread_t *thread;
 	pid_t p_pgid, c_pgid;
+#ifndef HAVE_SIGNALFD
+	sigset_t old_set, child_wait;
 
 	sigmask_func(0, NULL, &old_set);
 	if (!sigismember(&old_set, SIGCHLD)) {
@@ -388,17 +409,18 @@ script_killall(thread_master_t *m, int signo)
 		sigaddset(&child_wait, SIGCHLD);
 		sigmask_func(SIG_BLOCK, &child_wait, NULL);
 	}
-
-	thread = m->child.head;
+#endif
 
 	p_pgid = getpgid(0);
 
-	while (thread) {
+	for (thread = m->child.head; thread; thread = thread->next) {
 		c_pgid = getpgid(thread->u.c.pid);
-		if (c_pgid != p_pgid) {
+		if (c_pgid != p_pgid)
 			kill(-c_pgid, signo);
+		else {
+			log_message(LOG_INFO, "Child process %d in our process group %d", c_pgid, p_pgid);
+			kill(thread->u.c.pid, signo);
 		}
-		thread = thread->next;
 	}
 
 	if (!sigismember(&old_set, SIGCHLD))
