diff --git a/keepalived/check/check_misc.c b/keepalived/check/check_misc.c
index 77ea0b3c..174b5403 100644
--- a/keepalived/check/check_misc.c
+++ b/keepalived/check/check_misc.c
@@ -347,8 +347,13 @@ misc_check_child_thread(thread_t * thread)
 		if (timeout) {
 			/* If kill returns an error, we can't kill the process since either the process has terminated,
 			 * or we don't have permission. If we can't kill it, there is no point trying again. */
-			if (!kill(-pid, sig_num))
+			if (!kill(-pid, sig_num)) {
+				if (errno == ESRCH) {
+					/* The process has gone. Handle as termination */
+					// Set script_state to SCRIPT_STATE_DIED, and we should get a response on a wait
+				}
 				timeout = 1000;
+			}
 		} else if (misck_checker->state != SCRIPT_STATE_IDLE) {
 			log_message(LOG_INFO, "Child thread pid %d timeout with unknown script state %d", pid, misck_checker->state);
 			timeout = 10;	/* We need some timeout */
