diff --git a/lib/scheduler.c b/lib/scheduler.c
index 66102749..c9c9c527 100644
--- a/lib/scheduler.c
+++ b/lib/scheduler.c
@@ -238,9 +245,13 @@ report_child_status(int status, pid_t pid, char const *prog_name)
 	} else if (WIFSIGNALED(status)) {
 		if (WTERMSIG(status) == SIGSEGV) {
 			log_message(LOG_INFO, "%s exited due to segmentation fault (SIGSEGV).", prog_id);
-			log_message(LOG_INFO, "  Please report a bug at %s", "https://github.com/acassen/keepalived/issues");
-			log_message(LOG_INFO, "  %s", "and include this log from when keepalived started, what happened");
-			log_message(LOG_INFO, "  %s", "immediately before the crash, and your configuration file.");
+			log_message(LOG_INFO, "  %s", "Please report a bug at https://github.com/acassen/keepalived/issues");
+			log_message(LOG_INFO, "  %s", "and include this log from when keepalived started, a description");
+			log_message(LOG_INFO, "  %s", "of what happened before the crash and your configuration file.");
+			log_message(LOG_INFO, "  %s", "Also provide the output of keepalived -v, what Linux distro and version");
+			log_message(LOG_INFO, "  %s", "you are running on, and whether keepalived is being run in a container or VM.");
+			log_message(LOG_INFO, "  %s", "A failure to provide all this information may mean the crash cannot be investigated.");
+			log_message(LOG_INFO, "  %s", "If you are able to provide a stack backtrace with gdb that would really help.");
 		}
 		else
 			log_message(LOG_INFO, "%s exited due to signal %d", prog_id, WTERMSIG(status));
